src/main/java/com/aws/greengrass/mqttclient/MqttClient.java

```diff
@@ -433,50 +448,63 @@ public synchronized void unsubscribe(UnsubscribeRequest request)
     }
 
-    private synchronized void spoolMessage()  {
-        if (spoolingFuture.get() == null || spoolingFuture.get().isCancelled()) {
-            spoolingFuture.set(ses.scheduleWithFixedDelay(this::spoolTask, 0, 5, TimeUnit.SECONDS));
+    protected CompletableFuture<Integer> publishSingleTask() throws InterruptedException {
+        final long id = spool.popId();
+        PublishRequest request = spool.getMessageById(id);
+        if (request == null) {
+            return null;
         }
-    }
-
-    /**
-     * Iterate the spooler queue to publish all the spooled message.
-     */
-    protected void spoolTask() {
-        try {
-            getConnection(false).connect().get();
-            List<CompletableFuture<?>> publishRequests = new ArrayList<>();
-            while (!Thread.currentThread().isInterrupted() && mqttOnline.get() && spool.getCurrentMessageCount() > 0) {
-                final long id = spool.popId();
-                PublishRequest request = spool.getMessageById(id);
-                if (request == null) {
-                    continue;
-                }
-
-                MqttMessage m = new MqttMessage(request.getTopic(), request.getPayload());
-                publishRequests.add(getConnection(false).publish(m, request.getQos(), request.isRetain())
-                                .whenComplete((packetId, throwable) -> {
+        MqttMessage m = new MqttMessage(request.getTopic(), request.getPayload());
+        return getConnection(false).publish(m, request.getQos(), request.isRetain())
+                .whenComplete((packetId, throwable) -> {
                     // packetId is the SDK assigned ID. Ignore this and instead use the spooler ID
                     if (throwable == null) {
                         spool.removeMessageById(id);
                         logger.atDebug().kv("id", id).kv("topic", request.getTopic())
-                            .log("Successfully published message");
+                                .log("Successfully published message");
                     } else {
                         spool.addId(id);
                         logger.atError().log("Failed to publish the message via Spooler", throwable);
                     }
-                }));
+                });
+    }
 
-                if (publishRequests.size() >= maxInFlightPublishes) {
-                    CompletableFuture.anyOf(
-                            publishRequests.toArray(new CompletableFuture[0])).get();
-                    publishRequests = publishRequests.stream().filter(f -> !f.isDone()).collect(Collectors.toList());
+    /**
+     * Iterate the spooler queue to publish all the spooled message.
+     */
+    @SuppressWarnings("PMD.AvoidCatchingThrowable")
+    protected void publishAllTasks() {
+        List<CompletableFuture<?>> publishRequests = new ArrayList<>();
+        while (!Thread.currentThread().isInterrupted()) {
+            try {
+                getConnection(false).connect().get();
```

Comment:
Why are you manually calling connect? You shouldn't be doing this, the publish already handles it.

Reply:
Take this out of the loop. There's no reason for it to be inside the loop.

Reply:
In the edition of mainline, we called "getConnection(false).connect().get();" to make sure the OnInitConnect callback would be called and set the mqttClientOnline as True. So that the first run could start. 

The reason I put it in the while loop because the connect.get() may throw ExecutionExecption and the while loop may not start.